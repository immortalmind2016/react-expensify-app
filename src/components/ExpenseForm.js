import React from "react"
import moment from "moment" // deal with now time
import {SingleDatePicker} from "react-dates"

//const date=new Date();
const now=moment()
console.log(now.format('MMMM Do YYYY'))

export default class ExpenseForm extends React.Component{
    constructor(props){
        super(props);
        this.state={
        description:props.expense?props.expense.description:"",
        note:props.expense?props.expense.note:"",
        amount:props.expense?(props.expense.amount/100).toString():"",
        createdAt:props.expense?moment(props.expense.createdAt):moment(),
        calanderFocused:false,
        error:undefined,
        
    }
    
    }
    
     
   onDescriptionChange=(e)=>{
       const description=e.target.value;
        console.log(description)
       this.setState(()=>({description}))
 

   }
     onNoteChange=(e)=>{
       const note=e.target.value;
       this.setState(()=>({note}))
   }
   onAmountChange=(e)=>{
        const amount=e.target.value;
        if(!amount||amount.match(/^\d{1,}(\.\d{0,2})?$/)){this.setState(()=>({amount}))}
   } 
   onDateChange=(createdAt)=>{
       if(createdAt){
        this.setState({createdAt})
       }

   }
   onFocusChange=({focused})=>{
        this.setState({calanderFocused:focused})
   }
   onSubmit=(e)=>{
        e.preventDefault();
        if(!this.state.description||!this.state.amount){
            //set Error state equate to "please provide Description and amount "
            this.setState({error:"please provide Description and amount ."})
        }else{
            //cleare The error
            this.setState({error:undefined})
             this.props.onSubmit({
            description:this.state.description,
            amount:parseFloat(this.state.amount,10)*100,
            createdAt:this.state.createdAt.valueOf(),
            note:this.state.note
        })
        }   
       
   }
    render(){
       return (
      
          
             <form className="form" onSubmit={this.onSubmit}>
               {this.state.error &&<p className="form__error">{this.state.error}</p>}
              <input 
              className="text-input"
                type="text"
                placeholder="Description"
                autoFocus
                value={this.state.description}
                onChange={this.onDescriptionChange}
              />
              <input 
              className="text-input"
              type="text"
              placeholder="Amount"
              value={this.state.amount}
              onChange={this.onAmountChange}
                />
                 <div>
            <SingleDatePicker
            
              date={this.state.createdAt}
              onDateChange={this.onDateChange}
              focused={this.state.calanderFocused}
              onFocusChange={this.onFocusChange}
              numberOfMonths={1}
              isOutsideRange={(day)=>false}
            
             />
            </div>
              <textArea
              className="textarea"
              placeholder="add a note for your Expense (optional)"
            value={this.state.description}
            onChange={this.onNoteChange}
              >
              </textArea>
              <div>
               <button className="button-layout">{!this.props.expense?"Add Expense":"Save"}</button>
              
              </div>              
             

              
            
             </form>
      
       )
   }
}