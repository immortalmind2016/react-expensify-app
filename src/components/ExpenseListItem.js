import React from 'react'
import {connect} from "react-redux"
import {removeExpense,startRemoveExpense} from "../actions/expenses"
import {Link} from "react-router-dom"
import moment from "moment";
import numeral from "numeral"
// export a stateless functional Component
// description , amount , createdAt
// remove props.  ( called destructure)
export const ExpenseListItem=({dispatch,id,description , amount , createdAt})=>(

  
    <Link class="list-item" to={`/edit/${id}`}>
    <div>
    <h3 className="list-item__title"> {description}</h3>
     <span className="list-item__sub-title">
     {moment(createdAt).format("MMMM Do, YYYY")}</span>
    </div>
       <h3 className="list-item__data">
     {numeral(amount/100).format("$0,0.00")}
     </h3>

  
    </Link>
  
)
const mapStateToProps=(state)=>{
    return {
        filters:state.filters,
    
 
    }
}
export default connect(mapStateToProps)(ExpenseListItem);     

 

