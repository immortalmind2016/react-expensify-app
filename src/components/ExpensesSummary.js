import React from "react"
import {connect } from "react-redux"
import {Link} from "react-router-dom"
import selectExpenses from "../selectors/expenses"
import selectExpensesTotal from "../selectors/expenses-total"

import numeral from "numeral"

export const ExpensesSummary=(props)=>{
    const expenseWord=props.expensesCount===1?"expense":"expenses";
  return (
<div className="page-header">
    <div className="content-container">
   <h1 className="page-header__title">
    Viewing      <span> {props.expensesCount}  </span> {expenseWord}
      Total : <span>  {numeral(props.totalAmount/100).format("$0,0.00")}</span> 
       </h1>
       <div className="page-header__actions">
       <Link className="button-layout" to="/create">
       Add Expense 
       </Link>
       </div>
       
        </div>
    </div>
    )
}

const mapStateToProps=(state,props)=>{
     return {
         expensesCount:selectExpenses(state.expenses,state.filters).length,
         totalAmount:selectExpensesTotal(selectExpenses(state.expenses,state.filters))
     }
}

export default connect(mapStateToProps)(ExpensesSummary);