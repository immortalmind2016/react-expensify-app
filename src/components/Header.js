import React from "react"
import {NavLink} from "react-router-dom"
import {connect} from "react-redux"
import {startLogout} from "../actions/auth"
export const  Header=({startLogout})=>(
    <header>
        <h1>Expensify</h1>
        <div>
            <NavLink exact={true} to="/dashboard" activeClassName="is-active"> dashboard </NavLink>
            <NavLink to="/create" activeClassName="is-active"> Create </NavLink>
            <NavLink to="/edit/55" activeClassName="is-active"> edit </NavLink>
            <NavLink to="/help" activeClassName="is-active"> help </NavLink>
            <button onClick={startLogout}>Logout</button>
        </div>
    </header>
)

const mapDispatchToProps=(dispatch)=>({
    startLogout:()=>dispatch(startLogout())
})

export default connect(undefined,mapDispatchToProps)(Header)