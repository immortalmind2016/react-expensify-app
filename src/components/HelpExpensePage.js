import React from "react"

const HelpExpensePage=()=>(
    <div>
       this from my HelpExpensePage component
    </div>
)

export default HelpExpensePage