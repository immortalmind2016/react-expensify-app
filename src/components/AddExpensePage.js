import React from "react"
import ExpenseForm from "./ExpenseForm"
import {connect} from "react-redux"
import {startAddExpense,addExpense} from "../actions/expenses"

export class AddExpensePage extends React.Component{
    onSubmit=(expense)=>{
            //use push
            // attach then callback
            // dispatch action
            // redirect

        this.props.startAddExpense(expense)
           this.props.history.push("/") // redirect me with React 
         
            }
    render() {
        return (
        
        <div>
         <div className="page-header">
                <div className="content-container">
                      <h1 className="page-header__title">Add Expense</h1>
                </div>
            </div>
            <div className="content-container">
        
       <ExpenseForm 
       onSubmit={this.onSubmit}
       
       />
           </div>
    </div>
        )
    }
}

const mapDispatchToProps=(dispatch)=>({
    startAddExpense:(expense)=>dispatch(startAddExpense(expense))
})

export default connect(undefined,mapDispatchToProps)(AddExpensePage);