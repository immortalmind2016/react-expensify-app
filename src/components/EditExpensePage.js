import React from "react"
import {connect} from "react-redux"
import ExpenseForm from "./ExpenseForm"
import {history} from "../routers/AppRouter"
import {editExpense,setEditExpense,startRemoveExpense} from "../actions/expenses"
const EditExpensePage=(props)=>{
    return (
    <div >
        <div className="page-header">
         <div className="content-container">
         <h1 className="page-header__title">Edit expense</h1>
         </div>
        </div>
        <div className="content-container">
        <ExpenseForm 
            onSubmit={(expense)=>{
                console.log("updated",expense)
                props.dispatch(setEditExpense(props.match.params.id,expense))
                props.history.push("/")
            }}
            expense={props.expense}
        />
        <div style={{paddingBottom:"10px"}}>
       
        <button onClick={()=>{
            console.log(props.match.params)
       props.dispatch(startRemoveExpense({id:props.match.params.id}))
       history.push("/")
        }   }
         className="button-layout button--secondry">Remove Epense
         </button>
          </div>
        </div>
    </div>
    )
}
const mapStateToProps=(state,props)=>{
     return {
         expense:state.expenses.find((expense)=>(expense.id===props.match.params.id))
     }
}
export default connect(mapStateToProps)(EditExpensePage)