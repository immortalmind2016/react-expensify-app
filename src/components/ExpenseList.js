import React from "react"
import {connect} from "react-redux"
import selectExpenses from "../selectors/expenses"
import ExpenseListItem from "./ExpenseListItem"
// we added export named to Testing
export const ExpenseList=(props)=>(
    <div className="content-container">

        <div className="list-header">   
            <div className="show-for-mobile">Expenses</div>
            <div className="show-for-desktop">Expense</div>
            <div className="show-for-desktop">Amount</div>
        </div>
     <div className="list-body">

            {
                props.expenses.length===0?(
                    <div class="list-item list-item--message">
                    <span> No Expenses</span>
                    </div>
                    ):(
                    props.expenses.map((element)=>(
                    <ExpenseListItem key={element.id} {...element} />
                            ))
                )
            }
       </div>
    </div>
)

const mapStateToProps=(state)=>{
    return {
        expenses:selectExpenses(state.expenses,state.filters),
    }
}
export default connect(mapStateToProps)(ExpenseList);     
/*

const ConnectedExpensesList=connect((state)=>{
    return {
        expenses:state.expenses
    }
})(ExpenseList);     


 //have it's own HOC

export default ConnectedExpensesList

*/