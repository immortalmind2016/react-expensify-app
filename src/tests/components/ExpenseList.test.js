//react-test-render
//import ReactShallowRenderer from "react-test-renderer/shallow" we used Enzyme instead
import React from "react";
import {shallow} from "enzyme"
import {ExpenseList} from "../../components/ExpenseList"
const expenses=[
    {
        id:"1",
        description:"rent",
        amount:5,
        createdAt:0,
        note:"hh"
    }, {
        id:"2",
        description:"rent2",
        amount:50,
        createdAt:10,
        note:"hh"
    }
]
test("should render Expense list with expenses",()=>{
const wrapper=shallow(<ExpenseList expenses={expenses}/>);
//expect(wrapper.find("h1").text()).toBe("Expensify")
//expect(toJson(wrapper)).toMatchSnapshot();  
// now i can use it without toJson instead because i'm configurated my jest.config.json file
expect(wrapper).toMatchSnapshot();  
// snapMatchshoot byb2a 3ndi el Component shoot f file w yqarn

})

test("should render Expense list with Empty expenses array",()=>{

const wrapper=shallow(<ExpenseList expenses={[]}/>);

expect(wrapper).toMatchSnapshot();  

})



//test("should render header correctly ",()=>{
  //  const renderer=new ReactShallowRenderer();
   // renderer.render(<Header />)
   // expect(renderer.getRenderOutput()).toMatchSnapshot() 
    // will generate snapshot to my Component and save it
    // next time i can accept new changes or now
    // we will use enzyme because it's difcult to use renderer with complex components
    // raf Libirary -> request animation frame ( for browser )

//})