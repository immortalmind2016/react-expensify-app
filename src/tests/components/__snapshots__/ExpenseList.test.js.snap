// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Expense list with Empty expenses array 1`] = `
<div
  className="content-container"
>
  <div
    className="list-header"
  >
    <div
      className="show-for-mobile"
    >
      Expenses
    </div>
    <div
      className="show-for-desktop"
    >
      Expense
    </div>
    <div
      className="show-for-desktop"
    >
      Amount
    </div>
  </div>
  <div
    className="list-body"
  >
    <div
      class="list-item list-item--message"
    >
      <span>
         No Expenses
      </span>
    </div>
  </div>
</div>
`;

exports[`should render Expense list with expenses 1`] = `
<div
  className="content-container"
>
  <div
    className="list-header"
  >
    <div
      className="show-for-mobile"
    >
      Expenses
    </div>
    <div
      className="show-for-desktop"
    >
      Expense
    </div>
    <div
      className="show-for-desktop"
    >
      Amount
    </div>
  </div>
  <div
    className="list-body"
  >
    <Connect(ExpenseListItem)
      amount={5}
      createdAt={0}
      description="rent"
      id="1"
      key="1"
      note="hh"
    />
    <Connect(ExpenseListItem)
      amount={50}
      createdAt={10}
      description="rent2"
      id="2"
      key="2"
      note="hh"
    />
  </div>
</div>
`;
